### ISSUE TITLE
borg delete busy-waits when reading from remote

### ISSUE THREAD

**Issue by Voltara** [borgbackup/borg#940]
Running "borg delete" against a remote repository spends an inordinate amount of CPU time busy-waiting for input from the server.  According to strace, select immediately returns because file descriptor 5 is ready for writing, but the client then never actually writes anything.

    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})
    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})
    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})
    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})

    ... busy-waiting ...

    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})
    select(9, [6 8], [5], [5 6 8], {1, 0})  = 2 (in [6], out [5], left {0, 999999})
    mmap(NULL, 10489856, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f3272ecf000
    read(6, "\224\1\315\16\361\300\332\0274\2\1\0\371[\317\24E\260\344\3306\305C\246chunks\221\223"..., 10485760) = 5956
    mremap(0x7f3272ecf000, 10489856, 8192, MREMAP_MAYMOVE) = 0x7f3272ecf000
    munmap(0x7f32724ce000, 8192)            = 0
    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})
    write(5, "\224\1\315\17\261\246delete\221\332\0 \325/.y\370co\236\215H&E\233\302\300\361"..., 48) = 48
    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})
    write(5, "\224\1\315\17\262\243get\221\332\0 P\343\20\267\v\237\335:M\302\212\207\36\362\375\361\345\246\266"..., 45) = 45
    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})

    ... busy-waiting again ...

    select(9, [6 8], [5], [5 6 8], {1, 0})  = 1 (out [5], left {0, 999999})

**Comment by enkore** [comment]
16080fe fixes this for me (<10 % avg. CPU usage of the client instead of ~40 % when deleting remotely). I did some other testing, didn't seem to break anything or to run into timeouts (I set the timeout to 100 for testing).

**Comment by enkore** [comment]
That one is correct, though. When a delete is issued the command must be sent, so it select()s on the write fd once (unconditionally), and sends the command. Line 330 should then clear the write fd from the next wait. Hm, maybe this can be sped up by creating a delete_many method.

**Comment by Voltara** [comment]
I had a chance to dig a little bit deeper, and I'm starting to get an idea of what's happening (even with the new code.)

In a test run of `borg delete`, the busy-wait happened on a `get` command issued on a list of 22,285 IDs:

1. `call_many` starts by writing 100 calls to the server (100 `select` + 100 `write` of length 43; wouldn't it be better to combine these into a single 4300 byte write?)
2. At this point, there are 100 calls in-flight, and 22,185 remaining in the queue.
3. The loop is now stuck in a busy-waiting state until the server's response arrives: The `w_fds` list still contains `self.stdout_fd`, so that makes `select` return immediately.  No response data is available yet for reading, and the client is not yet willing to write (because it's at its limit of 100 in-flight requests) - so it keeps looping.

I think part of the issue is the function is calling `select` before it has decided what (if anything) it wants to write.

**Comment by enkore** [comment]
I revised the logic there a bit, so that first `to_send` is built and then the select is done - excluding the write fd if there is nothing to_send. However, that's only part of the issue. delete() is a "singular" API, meaning that (in my understanding) these requests will be issued one-by-one anyway. I guess that we'll want to queue/buffer deletes to send them in large batches.

**Comment by Voltara** [comment]
Steps to reproduce the behavior:
~~~
$ mkdir testdir && for i in `seq -w 0 99999`; do echo $i > testdir/$i; done
$ borg init -e none 127.0.100.1:/tmp/repo
$ borg create 127.0.100.1:/tmp/repo::test testdir
~~~

Simulate 1000ms of latency to 127.0.100.1 using Linux's netem feature:
~~~
$ sudo tc qdisc add dev lo root handle 1: prio bands 9 priomap 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8
$ sudo tc filter add dev lo protocol ip parent 1: prio 1 u32 match ip dst 127.0.100.1/32 flowid 1:1
$ sudo tc qdisc add dev lo parent 1:1 netem delay 1000ms
~~~

Before the fix (notice the user/system CPU time):
~~~
$ time borg check 127.0.100.1:/tmp/repo::test
Remote:
real    0m46.551s
user    0m9.660s
sys     0m2.476s
~~~

After the fix:
~~~
$ time borg check 127.0.100.1:/tmp/repo::test
Remote:
real    0m47.343s
user    0m2.184s
sys     0m0.128s
~~~

Undo the simulated latency:
~~~
$ sudo tc qdisc del dev lo parent 1:1 netem delay 1000ms
$ sudo tc filter del dev lo protocol ip parent 1: prio 1 u32 match ip dst 127.0.100.1/32 flowid 1:1
$ sudo tc qdisc del dev lo root handle 1: prio bands 9 priomap 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8
~~~

**Comment by ThomasWaldmann** [comment]
Numbers look much better, congrats!

**Comment by julian-klode** [comment]
Can we get this cherry-picked into 1.0-maint too? It's quite annoying there, and the patch applies cleanly.

**Comment by RonnyPfannschmidt** [comment]
@julian-klode just open a pr that way the maintainers just have to press the merge button with a smile

**Comment by julian-klode** [comment]
A pull request is in #1762

**Comment by ThomasWaldmann** [comment]
Reopening it for 1.0-maint potential adoption.

**Comment by ThomasWaldmann** [comment]
Fixed by #1762.
